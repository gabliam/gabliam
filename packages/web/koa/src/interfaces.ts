import { DefaultMethods } from '@gabliam/web-core';
import { koaRouter } from './koa';
/**
 * Represent a method that create an koa router
 */
export type RouterCreator = (path?: string) => koaRouter;

type Filter<T, U> = T extends U ? T : never;

export type KoaMethods = Filter<keyof koaRouter, DefaultMethods>;
