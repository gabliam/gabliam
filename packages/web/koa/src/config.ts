import { Bean, Joi, PluginConfig, Value } from '@gabliam/core';
import koaBody from 'koa-body';
import { KOA_PLUGIN_CONFIG } from './constants';

const koaBodyFormidableOptions = Joi.object()
  .keys({
    /**
     * {Integer} Limits the amount of memory all fields together (except files) can allocate in bytes.
     * If this value is exceeded, an 'error' event is emitted. The default size is 20MB.
     */
    maxFileSize: Joi.number().integer(),

    /**
     * {Integer} Limits the number of fields that the querystring parser will decode, default 1000
     */
    maxFields: Joi.number().integer(),

    /**
     * {Integer} Limits the amount of memory all fields together (except files) can allocate in bytes.
     * If this value is exceeded, an 'error' event is emitted, default 2mb (2 * 1024 * 1024)
     */
    maxFieldsSize: Joi.number().integer(),

    /**
     * {String} Sets the directory for placing file uploads in, default os.tmpDir()
     */
    uploadDir: Joi.string(),

    /**
     * {Boolean} Files written to uploadDir will include the extensions of the original files, default false
     */
    keepExtensions: Joi.bool(),

    /**
     * {String} If you want checksums calculated for incoming files, set this to either 'sha1' or 'md5', default false
     */
    hash: Joi.string(),

    /**
     * {Boolean} Multiple file uploads or no, default true
     */
    multiples: Joi.bool(),

    /**
     * {Function} Special callback on file begin. The function is executed directly by formidable.
     * It can be used to rename files before saving them to disk. See https://github.com/felixge/node-formidable#filebegin
     */
    onFileBegin: Joi.func(),
  })
  .default({ extended: false });

const koaBodyOptions = Joi.object().keys({
  /**
   * {Boolean} Patch request body to Node's ctx.req, default false
   *
   * Note: You can patch request body to Node or Koa in same time if you want.
   */
  patchNode: Joi.bool(),

  /**
   * {Boolean} Patch request body to Koa's ctx.request, default true
   *
   * Note: You can patch request body to Node or Koa in same time if you want.
   */
  patchKoa: Joi.bool(),

  /**
   * {String|Integer} The byte (if integer) limit of the JSON body, default 1mb
   */
  jsonLimit: [Joi.string(), Joi.number().integer()],

  /**
   * {String|Integer} The byte (if integer) limit of the form body, default 56kb
   */
  formLimit: [Joi.string(), Joi.number().integer()],

  /**
   * {String|Integer} The byte (if integer) limit of the text body, default 56kb
   */
  textLimit: [Joi.string(), Joi.number().integer()],

  /**
   * {String} Sets encoding for incoming form fields, default utf-8
   */
  encoding: Joi.string(),

  /**
   * {Boolean} Parse multipart bodies, default false
   */
  multipart: Joi.bool(),

  /**
   * {Boolean} Parse urlencoded bodies, default true
   */
  urlencoded: Joi.bool(),

  /**
   * {Boolean} Parse text bodies, default true
   */
  text: Joi.bool(),

  /**
   * {Boolean} Parse json bodies, default true
   */
  json: Joi.bool(),

  /**
   * Toggles co-body strict mode; if true, only parses arrays or objects, default true
   */
  jsonStrict: Joi.bool(),

  /**
   * {Object} Options to pass to the formidable multipart parser
   */
  formidable: koaBodyFormidableOptions,

  /**
   * {Function} Custom error handle, if throw an error, you can customize the response - onError(error, context), default will throw
   */
  onError: Joi.func(),

  /**
   * {Boolean} If enabled, don't parse GET, HEAD, DELETE requests, default true
   *
   * GET, HEAD, and DELETE requests have no defined semantics for the request body,
   * but this doesn't mean they may not be valid in certain use cases.
   * koa-body is strict by default
   *
   * see http://tools.ietf.org/html/draft-ietf-httpbis-p2-semantics-19#section-6.3
   */
  strict: Joi.bool(),
});

export interface KoaConfig {
  koaBodyOptions: koaBody.IKoaBodyOptions;
}

@PluginConfig()
export class KoaPluginConfig implements KoaConfig {
  @Value('application.web.koaBodyOptions', koaBodyOptions)
  koaBodyOptions: koaBody.IKoaBodyOptions;

  @Bean(KOA_PLUGIN_CONFIG)
  restConfig(): KoaConfig {
    return {
      koaBodyOptions: this.koaBodyOptions,
    };
  }
}
