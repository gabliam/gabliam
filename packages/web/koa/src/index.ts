import { KoaPlugin } from './koa-plugin';

export * from 'http-status-codes';
export * from './interfaces';
export * from './koa';
export * from './koa-converter';
export * from './koa-interceptor';
export * from './request';
export * from './response';
export { KoaPlugin as default };
