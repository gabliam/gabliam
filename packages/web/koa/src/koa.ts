import koa from 'koa';
import koaRouter from 'koa-router';

export { koa, koaRouter };
