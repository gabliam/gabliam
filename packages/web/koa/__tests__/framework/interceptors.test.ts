import {
  Config,
  Container,
  getMetadata,
  ReflectMetadata,
  Service,
} from '@gabliam/core';
import KoaPlugin, { koa, KoaConverter, toInterceptor } from '@gabliam/koa';
import {
  All,
  Controller,
  Delete,
  ExecContext,
  ExecutionContext,
  Get,
  Head,
  Interceptor,
  Next,
  Patch,
  Post,
  Put,
  UseInterceptors,
  WebConfig,
} from '@gabliam/web-core';
import { WebPluginTest } from '@gabliam/web-core/src/testing';
import { spy } from 'sinon';

let appTest: WebPluginTest;

beforeEach(async () => {
  appTest = new WebPluginTest([KoaPlugin]);
});

afterEach(async () => {
  await appTest.destroy();
});

describe('Complex interceptor', () => {
  let result: string;
  @Service()
  class A implements Interceptor {
    async intercept(@Next() next: () => Promise<any>) {
      result += 'a';
      await next();
      result += 'b';
    }
  }

  const spyA = spy(A.prototype, 'intercept');
  beforeEach(() => {
    result = '';
    spyA.resetHistory();
  });

  test('should call method-level interceptor correctly (GET)', async () => {
    @Controller('/')
    class TestController {
      @UseInterceptors(A)
      @Get('/')
      public getTest() {
        return 'GET';
      }
    }
    appTest.addClass(TestController);
    await appTest.buildAndStart();

    const response = await appTest
      .supertest()
      .get('/')
      .expect(200);

    expect(spyA.calledOnce).toBe(true);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
  });
});

describe('Interceptor:', () => {
  let result: string;

  @Service()
  class A implements Interceptor {
    async intercept(@Next() next: () => Promise<any>) {
      result += 'a';
      await next();
    }
  }

  async function b(ctx: any, nextFunc: () => Promise<any>) {
    result += 'b';
    await nextFunc();
  }

  const B = toInterceptor(b);

  @Service()
  class C implements Interceptor {
    async intercept() {
      result += 'c';
    }
  }

  const spyA = spy(A.prototype, 'intercept');
  const spyB = spy(B.prototype, 'intercept');
  const spyC = spy(C.prototype, 'intercept');

  beforeEach(() => {
    result = '';
    appTest.addClass(A);
    appTest.addClass(B);
    appTest.addClass(C);
    spyA.resetHistory();
    spyB.resetHistory();
    spyC.resetHistory();
  });

  test('should call method-level interceptor correctly (GET)', async () => {
    @Controller('/')
    class TestController {
      @UseInterceptors(A, B, C)
      @Get('/')
      public getTest() {
        return 'GET';
      }
    }
    appTest.addClass(TestController);
    await appTest.buildAndStart();

    const response = await appTest
      .supertest()
      .get('/')
      .expect(200);

    expect(spyA.calledOnce).toBe(true);
    expect(spyB.calledOnce).toBe(true);
    expect(spyC.calledOnce).toBe(true);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
  });

  test('should call method-level interceptor correctly (POST)', async () => {
    @Controller('/')
    class TestController {
      @UseInterceptors(A, B, C)
      @Post('/')
      public postTest() {
        return 'POST';
      }
    }

    appTest.addClass(TestController);
    await appTest.buildAndStart();

    const response = await appTest
      .supertest()
      .post('/')
      .expect(200);
    expect(spyA.calledOnce).toBe(true);
    expect(spyB.calledOnce).toBe(true);
    expect(spyC.calledOnce).toBe(true);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
  });

  test('should call method-level interceptor correctly (PUT)', async () => {
    @Controller('/')
    class TestController {
      @UseInterceptors(A, B, C)
      @Put('/')
      public postTest(ctx: koa.Context) {
        return 'PUT';
      }
    }
    appTest.addClass(TestController);
    await appTest.buildAndStart();

    const response = await appTest
      .supertest()
      .put('/')
      .expect(200);
    expect(spyA.calledOnce).toBe(true);
    expect(spyB.calledOnce).toBe(true);
    expect(spyC.calledOnce).toBe(true);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
  });

  test('should call method-level interceptor correctly (PATCH)', async () => {
    @Controller('/')
    class TestController {
      @UseInterceptors(A, B, C)
      @Patch('/')
      public postTest(ctx: koa.Context) {
        return 'PATCH';
      }
    }

    appTest.addClass(TestController);
    await appTest.buildAndStart();

    const response = await appTest
      .supertest()
      .patch('/')
      .expect(200);
    expect(spyA.calledOnce).toBe(true);
    expect(spyB.calledOnce).toBe(true);
    expect(spyC.calledOnce).toBe(true);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
  });

  test('should call method-level interceptor correctly (HEAD)', async () => {
    @Controller('/')
    class TestController {
      @UseInterceptors(A, B, C)
      @Head('/')
      public postTest() {
        return 'HEAD';
      }
    }
    appTest.addClass(TestController);
    await appTest.buildAndStart();

    const response = await appTest
      .supertest()
      .head('/')
      .expect(200);
    expect(spyA.calledOnce).toBe(true);
    expect(spyB.calledOnce).toBe(true);
    expect(spyC.calledOnce).toBe(true);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
  });

  test('should call method-level interceptor correctly (DELETE)', async () => {
    @Controller('/')
    class TestController {
      @UseInterceptors(A, B, C)
      @Delete('/')
      public postTest() {
        return 'DELETE';
      }
    }
    appTest.addClass(TestController);
    await appTest.buildAndStart();

    const response = await appTest
      .supertest()
      .delete('/')
      .expect(200);
    expect(spyA.calledOnce).toBe(true);
    expect(spyB.calledOnce).toBe(true);
    expect(spyC.calledOnce).toBe(true);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
  });

  test('should call method-level interceptor correctly (ALL)', async () => {
    @Controller('/')
    class TestController {
      @UseInterceptors(A, B, C)
      @All('/')
      public postTest() {
        return 'ALL';
      }
    }
    appTest.addClass(TestController);
    await appTest.buildAndStart();

    const response = await appTest
      .supertest()
      .get('/')
      .expect(200);
    expect(spyA.calledOnce).toBe(true);
    expect(spyB.calledOnce).toBe(true);
    expect(spyC.calledOnce).toBe(true);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
  });

  test('should call controller-level interceptor correctly', async () => {
    @UseInterceptors(A, B, C)
    @Controller({
      path: '/',
    })
    class TestController {
      @Get('/')
      public getTest() {
        return 'GET';
      }
    }

    appTest.addClass(TestController);
    await appTest.buildAndStart();

    const response = await appTest
      .supertest()
      .get('/')
      .expect(200);
    expect(spyA.calledOnce).toBe(true);
    expect(spyB.calledOnce).toBe(true);
    expect(spyC.calledOnce).toBe(true);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
  });

  test('should call server-level interceptor correctly', async () => {
    @Controller('/')
    class TestController {
      @Get('/')
      public getTest() {
        return 'GET';
      }
    }

    @Config()
    class ServerConfig {
      @WebConfig()
      serverConfig(app: koa, container: Container) {
        const converter = container.get(KoaConverter);
        app.use(converter.interceptorToMiddleware(A));
        app.use(b);
        app.use(converter.interceptorToMiddleware(C));
      }
    }

    appTest.addClass(TestController);
    appTest.addClass(ServerConfig);
    await appTest.buildAndStart();

    const response = await appTest
      .supertest()
      .get('/')
      .expect(200);
    expect(spyA.calledOnce).toBe(true);
    expect(spyC.calledOnce).toBe(true);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
  });

  test('should call all interceptor in correct order', async () => {
    @UseInterceptors(B)
    @Controller({
      path: '/',
    })
    class TestController {
      @UseInterceptors(C)
      @Get('/')
      public getTest() {
        return 'GET';
      }
    }

    @Config()
    class ServerConfig {
      @WebConfig()
      serverConfig(app: koa, container: Container) {
        const converter = container.get(KoaConverter);
        app.use(converter.interceptorToMiddleware(A));
      }
    }

    appTest.addClass(TestController);
    appTest.addClass(ServerConfig);
    await appTest.buildAndStart();

    const response = await appTest
      .supertest()
      .get('/')
      .expect(200);
    expect(spyA.calledOnce).toBe(true);
    expect(spyB.calledOnce).toBe(true);
    expect(spyC.calledOnce).toBe(true);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
  });

  test('should resolve controller-level interceptor', async () => {
    const symbolId = Symbol('spyA');
    const strId = 'spyB';

    @UseInterceptors(symbolId, strId)
    @Controller({
      path: '/',
    })
    class TestController {
      @Get('/')
      public getTest() {
        return 'GET';
      }
    }

    appTest.addClass(TestController);
    appTest.gab.container
      .bind(symbolId)
      .to(A)
      .inSingletonScope();
    appTest.gab.container
      .bind(strId)
      .to(B)
      .inSingletonScope();
    await appTest.buildAndStart();

    const response = await appTest
      .supertest()
      .get('/')
      .expect(200);
    expect(spyA.calledOnce).toBe(true);
    expect(spyB.calledOnce).toBe(true);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
  });

  test('should resolve method-level interceptor', async () => {
    const symbolId = Symbol('spyA');
    const strId = 'spyB';

    @Controller('/')
    class TestController {
      @UseInterceptors(symbolId, strId)
      @Get('/')
      public getTest() {
        return 'GET';
      }
    }

    appTest.addClass(TestController);
    appTest.gab.container
      .bind(symbolId)
      .to(A)
      .inSingletonScope();
    appTest.gab.container
      .bind(strId)
      .to(B)
      .inSingletonScope();
    await appTest.buildAndStart();

    const response = await appTest
      .supertest()
      .get('/')
      .expect(200);
    expect(spyA.calledOnce).toBe(true);
    expect(spyB.calledOnce).toBe(true);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
  });

  test('should compose controller- and method-level interceptor', async () => {
    const symbolId = Symbol('spyA');
    const strId = 'spyB';

    @UseInterceptors(symbolId)
    @Controller({
      path: '/',
    })
    class TestController {
      @UseInterceptors(strId)
      @Get('/')
      public getTest() {
        return 'GET';
      }
    }

    appTest.addClass(TestController);
    appTest.gab.container
      .bind(symbolId)
      .to(A)
      .inSingletonScope();
    appTest.gab.container
      .bind(strId)
      .to(B)
      .inSingletonScope();
    await appTest.buildAndStart();

    const response = await appTest
      .supertest()
      .get('/')
      .expect(200);
    expect(spyA.calledOnce).toBe(true);
    expect(spyB.calledOnce).toBe(true);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
  });
});

describe('Interceptor inject:', () => {
  let result: string;
  let args: string;
  @Service()
  class A implements Interceptor {
    intercept() {
      result += 'a';
    }
  }

  @Service()
  class B implements Interceptor {
    intercept() {
      result += 'b';
    }
  }

  @Service()
  class C implements Interceptor {
    intercept(@ExecContext() execCtx: ExecutionContext) {
      result += 'c';
      const argsCtx = getMetadata<string[]>('args', execCtx.getHandler());
      if (argsCtx) {
        args += argsCtx;
      }
    }
  }

  @Service()
  class D implements Interceptor {
    intercept(@ExecContext() execCtx: ExecutionContext) {
      result += 'd';
      const argsCtx = getMetadata<string[]>('args2', execCtx.getHandler());
      if (argsCtx) {
        args += argsCtx.join('');
      }
    }
  }

  const AddArgs = (...s: string[]) => ReflectMetadata('args', s);
  const AddArgs2 = (...s: string[]) => ReflectMetadata('args2', s);

  beforeEach(() => {
    result = '';
    args = '';
    appTest
      .addClass(A)
      .addClass(B)
      .addClass(C)
      .addClass(D);
  });

  test('should call method-level interceptor correctly (GET)', async () => {
    @Controller('/')
    class TestController {
      @UseInterceptors(A, B, C, D)
      @AddArgs('carg')
      @AddArgs2('dearg', 'dearg2')
      @Get('/')
      public getTest() {
        return 'GET';
      }
    }

    appTest.addClass(TestController);
    await appTest.buildAndStart();

    const response = await appTest
      .supertest()
      .get('/')
      .expect(200);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
    expect(args).toMatchSnapshot();
  });

  test('should call method-level interceptor correctly (POST)', async () => {
    @Controller('/')
    class TestController {
      @UseInterceptors(A, B, C, D)
      @AddArgs('carg')
      @AddArgs2('dearg', 'dearg2')
      @Post('/')
      public postTest() {
        return 'POST';
      }
    }

    appTest.addClass(TestController);
    await appTest.buildAndStart();

    const response = await appTest
      .supertest()
      .post('/')
      .expect(200);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
    expect(args).toMatchSnapshot();
  });

  test('should call method-level interceptor correctly (PUT)', async () => {
    @Controller('/')
    class TestController {
      @UseInterceptors(A, B, C, D)
      @AddArgs('carg')
      @AddArgs2('dearg', 'dearg2')
      @Put('/')
      public postTest() {
        return 'PUT';
      }
    }

    appTest.addClass(TestController);
    await appTest.buildAndStart();

    const response = await appTest
      .supertest()
      .put('/')
      .expect(200);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
    expect(args).toMatchSnapshot();
  });

  test('should call method-level interceptor correctly (PATCH)', async () => {
    @Controller('/')
    class TestController {
      @UseInterceptors(A, B, C, D)
      @AddArgs('carg')
      @AddArgs2('dearg', 'dearg2')
      @Patch('/')
      public postTest() {
        return 'PATCH';
      }
    }

    appTest.addClass(TestController);
    await appTest.buildAndStart();

    const response = await appTest
      .supertest()
      .patch('/')
      .expect(200);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
    expect(args).toMatchSnapshot();
  });

  test('should call method-level interceptor correctly (HEAD)', async () => {
    @Controller('/')
    class TestController {
      @UseInterceptors(A, B, C, D)
      @AddArgs('carg')
      @AddArgs2('dearg', 'dearg2')
      @Head('/')
      public postTest() {
        return 'HEAD';
      }
    }
    appTest.addClass(TestController);
    await appTest.buildAndStart();

    const response = await appTest
      .supertest()
      .head('/')
      .expect(200);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
    expect(args).toMatchSnapshot();
  });

  test('should call method-level interceptor correctly (DELETE)', async () => {
    @Controller('/')
    class TestController {
      @UseInterceptors(A, B, C, D)
      @AddArgs('carg')
      @AddArgs2('dearg', 'dearg2')
      @Delete('/')
      public postTest() {
        return 'DELETE';
      }
    }

    appTest.addClass(TestController);
    await appTest.buildAndStart();

    const response = await appTest
      .supertest()
      .delete('/')
      .expect(200);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
    expect(args).toMatchSnapshot();
  });

  test('should call method-level interceptor correctly (ALL)', async () => {
    @Controller('/')
    class TestController {
      @UseInterceptors(A, B, C, D)
      @AddArgs('carg')
      @AddArgs2('dearg', 'dearg2')
      @All('/')
      public postTest() {
        return 'ALL';
      }
    }

    appTest.addClass(TestController);
    await appTest.buildAndStart();

    const response = await appTest
      .supertest()
      .get('/')
      .expect(200);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
    expect(args).toMatchSnapshot();
  });

  test('should call controller-level interceptor correctly', async () => {
    @UseInterceptors(A, B, C, D)
    @AddArgs('carg')
    @AddArgs2('dearg', 'dearg2')
    @Controller({
      path: '/',
    })
    class TestController {
      @Get('/')
      public getTest() {
        return 'GET';
      }
    }

    appTest.addClass(TestController);
    await appTest.buildAndStart();

    const response = await appTest
      .supertest()
      .get('/')
      .expect(200);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
    expect(args).toMatchSnapshot();
  });

  test('should call all interceptor in correct order', async () => {
    @UseInterceptors(A, B)
    @Controller({
      path: '/',
    })
    class TestController {
      @UseInterceptors(C, D)
      @AddArgs('carg')
      @AddArgs2('dearg', 'dearg2')
      @Get('/')
      public getTest() {
        return 'GET';
      }
    }

    appTest.addClass(TestController);
    await appTest.buildAndStart();

    const response = await appTest
      .supertest()
      .get('/')
      .expect(200);
    expect(response).toMatchSnapshot();
    expect(result).toMatchSnapshot();
    expect(args).toMatchSnapshot();
  });
});
