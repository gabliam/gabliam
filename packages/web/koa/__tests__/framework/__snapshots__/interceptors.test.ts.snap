// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Complex interceptor should call method-level interceptor correctly (GET) 1`] = `
{
  "header": {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": {
    "data": undefined,
    "headers": {},
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`Complex interceptor should call method-level interceptor correctly (GET) 2`] = `"ab"`;

exports[`Interceptor inject: should call all interceptor in correct order 1`] = `
{
  "header": {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": {
    "data": undefined,
    "headers": {},
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`Interceptor inject: should call all interceptor in correct order 2`] = `"abcd"`;

exports[`Interceptor inject: should call all interceptor in correct order 3`] = `"cargdeargdearg2"`;

exports[`Interceptor inject: should call controller-level interceptor correctly 1`] = `
{
  "header": {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": {
    "data": undefined,
    "headers": {},
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`Interceptor inject: should call controller-level interceptor correctly 2`] = `"abcd"`;

exports[`Interceptor inject: should call controller-level interceptor correctly 3`] = `""`;

exports[`Interceptor inject: should call method-level interceptor correctly (ALL) 1`] = `
{
  "header": {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": {
    "data": undefined,
    "headers": {},
    "method": "GET",
  },
  "status": 200,
  "text": "ALL",
}
`;

exports[`Interceptor inject: should call method-level interceptor correctly (ALL) 2`] = `"abcd"`;

exports[`Interceptor inject: should call method-level interceptor correctly (ALL) 3`] = `"cargdeargdearg2"`;

exports[`Interceptor inject: should call method-level interceptor correctly (DELETE) 1`] = `
{
  "header": {
    "connection": "close",
    "content-length": "6",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": {
    "data": undefined,
    "headers": {},
    "method": "DELETE",
  },
  "status": 200,
  "text": "DELETE",
}
`;

exports[`Interceptor inject: should call method-level interceptor correctly (DELETE) 2`] = `"abcd"`;

exports[`Interceptor inject: should call method-level interceptor correctly (DELETE) 3`] = `"cargdeargdearg2"`;

exports[`Interceptor inject: should call method-level interceptor correctly (GET) 1`] = `
{
  "header": {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": {
    "data": undefined,
    "headers": {},
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`Interceptor inject: should call method-level interceptor correctly (GET) 2`] = `"abcd"`;

exports[`Interceptor inject: should call method-level interceptor correctly (GET) 3`] = `"cargdeargdearg2"`;

exports[`Interceptor inject: should call method-level interceptor correctly (HEAD) 1`] = `
{
  "header": {
    "connection": "close",
    "content-length": "4",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": {
    "data": undefined,
    "headers": {},
    "method": "HEAD",
  },
  "status": 200,
  "text": undefined,
}
`;

exports[`Interceptor inject: should call method-level interceptor correctly (HEAD) 2`] = `"abcd"`;

exports[`Interceptor inject: should call method-level interceptor correctly (HEAD) 3`] = `"cargdeargdearg2"`;

exports[`Interceptor inject: should call method-level interceptor correctly (PATCH) 1`] = `
{
  "header": {
    "connection": "close",
    "content-length": "5",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": {
    "data": undefined,
    "headers": {},
    "method": "PATCH",
  },
  "status": 200,
  "text": "PATCH",
}
`;

exports[`Interceptor inject: should call method-level interceptor correctly (PATCH) 2`] = `"abcd"`;

exports[`Interceptor inject: should call method-level interceptor correctly (PATCH) 3`] = `"cargdeargdearg2"`;

exports[`Interceptor inject: should call method-level interceptor correctly (POST) 1`] = `
{
  "header": {
    "connection": "close",
    "content-length": "4",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": {
    "data": undefined,
    "headers": {},
    "method": "POST",
  },
  "status": 200,
  "text": "POST",
}
`;

exports[`Interceptor inject: should call method-level interceptor correctly (POST) 2`] = `"abcd"`;

exports[`Interceptor inject: should call method-level interceptor correctly (POST) 3`] = `"cargdeargdearg2"`;

exports[`Interceptor inject: should call method-level interceptor correctly (PUT) 1`] = `
{
  "header": {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": {
    "data": undefined,
    "headers": {},
    "method": "PUT",
  },
  "status": 200,
  "text": "PUT",
}
`;

exports[`Interceptor inject: should call method-level interceptor correctly (PUT) 2`] = `"abcd"`;

exports[`Interceptor inject: should call method-level interceptor correctly (PUT) 3`] = `"cargdeargdearg2"`;

exports[`Interceptor: should call all interceptor in correct order 1`] = `
{
  "header": {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": {
    "data": undefined,
    "headers": {},
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`Interceptor: should call all interceptor in correct order 2`] = `"abc"`;

exports[`Interceptor: should call controller-level interceptor correctly 1`] = `
{
  "header": {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": {
    "data": undefined,
    "headers": {},
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`Interceptor: should call controller-level interceptor correctly 2`] = `"abc"`;

exports[`Interceptor: should call method-level interceptor correctly (ALL) 1`] = `
{
  "header": {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": {
    "data": undefined,
    "headers": {},
    "method": "GET",
  },
  "status": 200,
  "text": "ALL",
}
`;

exports[`Interceptor: should call method-level interceptor correctly (ALL) 2`] = `"abc"`;

exports[`Interceptor: should call method-level interceptor correctly (DELETE) 1`] = `
{
  "header": {
    "connection": "close",
    "content-length": "6",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": {
    "data": undefined,
    "headers": {},
    "method": "DELETE",
  },
  "status": 200,
  "text": "DELETE",
}
`;

exports[`Interceptor: should call method-level interceptor correctly (DELETE) 2`] = `"abc"`;

exports[`Interceptor: should call method-level interceptor correctly (GET) 1`] = `
{
  "header": {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": {
    "data": undefined,
    "headers": {},
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`Interceptor: should call method-level interceptor correctly (GET) 2`] = `"abc"`;

exports[`Interceptor: should call method-level interceptor correctly (HEAD) 1`] = `
{
  "header": {
    "connection": "close",
    "content-length": "4",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": {
    "data": undefined,
    "headers": {},
    "method": "HEAD",
  },
  "status": 200,
  "text": undefined,
}
`;

exports[`Interceptor: should call method-level interceptor correctly (HEAD) 2`] = `"abc"`;

exports[`Interceptor: should call method-level interceptor correctly (PATCH) 1`] = `
{
  "header": {
    "connection": "close",
    "content-length": "5",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": {
    "data": undefined,
    "headers": {},
    "method": "PATCH",
  },
  "status": 200,
  "text": "PATCH",
}
`;

exports[`Interceptor: should call method-level interceptor correctly (PATCH) 2`] = `"abc"`;

exports[`Interceptor: should call method-level interceptor correctly (POST) 1`] = `
{
  "header": {
    "connection": "close",
    "content-length": "4",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": {
    "data": undefined,
    "headers": {},
    "method": "POST",
  },
  "status": 200,
  "text": "POST",
}
`;

exports[`Interceptor: should call method-level interceptor correctly (POST) 2`] = `"abc"`;

exports[`Interceptor: should call method-level interceptor correctly (PUT) 1`] = `
{
  "header": {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": {
    "data": undefined,
    "headers": {},
    "method": "PUT",
  },
  "status": 200,
  "text": "PUT",
}
`;

exports[`Interceptor: should call method-level interceptor correctly (PUT) 2`] = `"abc"`;

exports[`Interceptor: should call server-level interceptor correctly 1`] = `
{
  "header": {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": {
    "data": undefined,
    "headers": {},
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`Interceptor: should call server-level interceptor correctly 2`] = `"abc"`;

exports[`Interceptor: should compose controller- and method-level interceptor 1`] = `
{
  "header": {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": {
    "data": undefined,
    "headers": {},
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`Interceptor: should compose controller- and method-level interceptor 2`] = `"ab"`;

exports[`Interceptor: should resolve controller-level interceptor 1`] = `
{
  "header": {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": {
    "data": undefined,
    "headers": {},
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`Interceptor: should resolve controller-level interceptor 2`] = `"ab"`;

exports[`Interceptor: should resolve method-level interceptor 1`] = `
{
  "header": {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/plain; charset=utf-8",
  },
  "req": {
    "data": undefined,
    "headers": {},
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`Interceptor: should resolve method-level interceptor 2`] = `"ab"`;
